From 342162d831a058b2f81816e3bf92e350bb1a9943 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?S=C3=B6ren=20Tempel?= <soeren+git@soeren-tempel.net>
Date: Sat, 16 Oct 2021 23:47:01 +0200
Subject: [PATCH] Fix compilation with libusb < 1.0.24

With libusb 1.0.24 the function parameter for libusb_strerror was
changed from `enum libusb_error` to `int` [0]. However, some Linux
distributions (e.g. Ubuntu 20.04) still ship a libusb version < 1.0.24
which causes this code to not compile since C++ allows implicit
conversions from enum to int but not from int to enum [1]. To fix this
issue, cast the value passed to libusb_strerror explicitly to an enum
which should cause an implicit conversion from enum to int on Linux
distributions with libusb >= 1.0.24.

See https://github.com/nmeum/android-tools/pull/48#issuecomment-944891603

[0]: https://github.com/libusb/libusb/commit/5d0d0a58ed3ea813200b83d69e28a3ea24e86f01
[1]: https://en.cppreference.com/w/cpp/language/enum
---
 client/usb_libusb.cpp | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/client/usb_libusb.cpp b/client/usb_libusb.cpp
index 71b077cc..3a711795 100644
--- a/client/usb_libusb.cpp
+++ b/client/usb_libusb.cpp
@@ -294,7 +294,7 @@ struct LibusbConnection : public Connection {
         read->active = true;
         int rc = libusb_submit_transfer(read->transfer);
         if (rc != 0) {
-            LOG(ERROR) << "libusb_submit_transfer failed: " << libusb_strerror(rc);
+            LOG(ERROR) << "libusb_submit_transfer failed: " << libusb_strerror(static_cast<libusb_error>(rc));
         }
     }
 
@@ -495,7 +495,7 @@ struct LibusbConnection : public Connection {
         int rc = libusb_open(device_.get(), &handle_raw);
         if (rc != 0) {
             // TODO: Handle no permissions.
-            LOG_ERR(error, "failed to open device: %s", libusb_strerror(rc));
+            LOG_ERR(error, "failed to open device: %s", libusb_strerror(static_cast<libusb_error>(rc)));
             return false;
         }
 
